cmake: mcuboot: Use imgtool instead of west for signing

nordicjm opened the pull request:
modules: mcuboot: Add Kconfig for firmware updater image
    
    Adds a Kconfig which can be selected when building for firmware
    updater mode to select if this is the application or the firmware
    updater image that is being built



    cmake: mcuboot: Use imgtool instead of west for signing
    
    Uses imgtool directly to sign images instead of calling west sign,
    this also removes the MCUBOOT_CMAKE_WEST_SIGN_PARAMS Kconfig option
    as this has no effect



    doc: release: 3.4: Remove Kconfig option usage for removed Kconfig
    
    Removes the Kconfig extension function around a Kconfig that has
    been removed



    scripts: west_commands: sign: Deprecate imgtool signing
    
    Deprecates signing for imgtool using west sign



    doc: release: 4.0: Add note on imgtool west sign change
    
    Adds a note that the build system now uses imgtool directly



    doc: migration: 4.0: Add notes on deprecation of imgtool west sign
    
    Adds details on the deprecation of this feature


Fixes #78044

nordicjm committed:
modules: mcuboot: Add Kconfig for firmware updater image

Adds a Kconfig which can be selected when building for firmware
updater mode to select if this is the application or the firmware
updater image that is being built

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
cmake: mcuboot: Use imgtool instead of west for signing

Uses imgtool directly to sign images instead of calling west sign,
this also removes the MCUBOOT_CMAKE_WEST_SIGN_PARAMS Kconfig option
as this has no effect

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
doc: release: 3.4: Remove Kconfig option usage for removed Kconfig

Removes the Kconfig extension function around a Kconfig that has
been removed

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
scripts: west_commands: sign: Deprecate imgtool signing

Deprecates signing for imgtool using west sign

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
doc: release: 4.0: Add note on imgtool west sign change

Adds a note that the build system now uses imgtool directly

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
doc: migration: 4.0: Add notes on deprecation of imgtool west sign

Adds details on the deprecation of this feature

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

pdgendt commented on modules/Kconfig.mcuboot:
```diff
@@ -256,6 +244,16 @@ config MCUBOOT_BOOTLOADER_NO_DOWNGRADE
+	bool "Application is firmware updater image"
+	depends on MCUBOOT_BOOTLOADER_MODE_FIRMWARE_UPDATER
+	help
+	  Select this is the current image is the firmware updater image. This will use slot 1
+	  information when signing the image.
```
```suggestion
	  Select this if the current image is the firmware updater image. This will use slot 1
	  information when signing the image.
```

nordicjm replied:
Fixed

nordicjm committed:
doc: release: 3.4: Use double backticks

Fixes a compliance error

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

mmahadevan108 commented:
cc @butok

butok commented:
The Zephyr "west sign" documentation should be updated https://docs.zephyrproject.org/latest/develop/west/sign.html

nordicjm committed:
doc: build: Add signing page

Adds a signing page under build, which contains information on
how to sign images from the build system, this removes the old
details for using west sign with imgtool from the west sign page

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm commented:
> The Zephyr "west sign" documentation should be updated https://docs.zephyrproject.org/latest/develop/west/sign.html

Updated

mmahadevan108 commented:
@tejlmand , kindly review.

tejlmand commented on modules/Kconfig.mcuboot:
```diff
@@ -256,6 +256,16 @@ config MCUBOOT_BOOTLOADER_NO_DOWNGRADE
+	  Select this if the current image is the firmware updater image. This will use slot 1
+	  information when signing the image.
+
+	  Note that the zephyr chosen node ``zephyr,code-partition`` should be set to
+	  ``slot1_partition`` for this image.
```
if `this image` is the image currently being built, then isn't it possible to add a dependency to the devicetree chosen node ?

tejlmand replied:
discussed offline.
I believe the help text can be improved, but that is just nice to have.

tejlmand commented on cmake/mcuboot.cmake:
```diff
@@ -73,12 +73,25 @@ function(zephyr_mcuboot_tasks)
-    --build-dir "${APPLICATION_BINARY_DIR}")
+  # Fetch devicetree details for flash and slot information
+  dt_chosen(flash_node PROPERTY "zephyr,flash")
+  dt_nodelabel(slot0_flash NODELABEL "slot0_partition")
```
should this have the `REQUIRED` flag ?

I don't see any handling of the case where no node exists with nodelabel `"slot0_partition"`.

tejlmand commented on cmake/mcuboot.cmake:
```diff
@@ -73,12 +73,25 @@ function(zephyr_mcuboot_tasks)
+  if(NOT CONFIG_MCUBOOT_BOOTLOADER_MODE_SINGLE_APP AND (NOT CONFIG_MCUBOOT_BOOTLOADER_MODE_FIRMWARE_UPDATER OR CONFIG_MCUBOOT_APPLICATION_FIRMWARE_UPDATER))
+    # Slot 1 size is used instead of slot 0 size
+    set(slot_size)
+    dt_nodelabel(slot1_flash NODELABEL "slot1_partition")
```
ditto here, has it been guranteed elsewhere that a node with nodelabel `slot1_partition` does exists ?

tejlmand reviewed (changes requested):
some observations related to the use of devicetree CMake functions which should be addressed or replied if this is handled / guaranteed elsewhere in the build system.

Rest looks fine.

tejlmand commented on cmake/mcuboot.cmake:
```diff
@@ -73,12 +73,30 @@ function(zephyr_mcuboot_tasks)
+
+  # If single slot mode, or if in firmware updater mode and this is the firmware updater image,
+  # use slot 0 information
+  if(NOT CONFIG_MCUBOOT_BOOTLOADER_MODE_SINGLE_APP AND (NOT CONFIG_MCUBOOT_BOOTLOADER_MODE_FIRMWARE_UPDATER OR CONFIG_MCUBOOT_APPLICATION_FIRMWARE_UPDATER))
```
minor nit, quite a long line.

d3zd3z reviewed (approved):
I do think this is overall a good thing. We should probably get some feedback about deprecating the west command, although I'm not sure it is all that beneficial, since imgtool has to be installed.

marc-hb commented:
Please cc: some "rimage" person next time: I found this by sheer chance thanks to a random git conflict in sign.py!

`imgtool` and `rimage` had little in common and I'm not convinced the `sign.py` "abstraction" ever made a lot of sense but from an `rimage` and SOF perspective it's been extremely useful to avoid... coding in CMake, which means anyone can and did make changes.

Also note: `rimage` does MORE than just sign SOF, see for instance https://github.com/thesofproject/sof/commits/b107e40976e33/tools/rimage/config

cc: @kv2019i , @lyakh , @dbaluta , @LaurentiuM1234 , @iuliana-prodan

nordicjm commented:
west sign is not being deprecated, it is just the imgtool part of west sign, there are no plans to deprecate or remove anything with rimage, the imgtool part was because it was duplicating parts in different areas and the west sign part was missing parts

marc-hb commented:
Thanks @nordicjm for the clarification! Still, please make sure some "rimage person" is copied on future `sign.py` changes anyway. @kv2019i, @lyak, @dbaluta please make the corresponding MAINTAINERS.yml change?

Additional cc: @teburd and @pillo79 for LLEXT signing.

teburd commented:
I'd still recommend using build targets generated by cmake for this sort of thing myself, but to each their own.

marc-hb commented:
Please don't get me wrong: anything that looks like a _dependency graph_ should be in CMake, not in Python. The right tool for the job! But because CMake is so central, it may seem "easier" to do everything in CMake if you're one of the lucky(?) few who understand it and _that_ would be bad.

Take for instance a 30 seconds look at `RimageSigner.sign()` as of today's commit:

https://github.com/zephyrproject-rtos/zephyr/blob/10d49736cffa14d3798e615e70d58b3be8ee2cfc/scripts/west_commands/sign.py#L466

It's not pretty but imagine for a moment coding the same logic in CMake! As a job security attempt maybe?

I don't think anyone would seriously consider re-implementing `scripts/build/gen_kobject_list.py` or one of its friends in CMake either?
